clap_complete/src/shells/shell.rs

```diff
@@ -89,3 +90,39 @@ impl Generator for Shell {
         }
     }
 }
+
+impl Shell {
+    /// Determine the user's current shell from the environment
+    ///
+    /// This will read the SHELL environment variable and try to determine which shell is in use
+    /// from that.
+    ///
+    /// If SHELL is not set, then on windows, it will default to powershell, and on
+    /// other OSes it will return an error.
+    ///
+    /// If SHELL is set, but contains a value that doesn't correspond to one of the supported shell
+    /// types, then an error is returned.
+    ///
+    /// # Example:
+    ///
+    /// ```ignore
+    /// use clap_complete::{generate, shells::Shell}
+    /// let mut cmd = build_cli();
+    /// generate(Shell::from_env().unwrap_or(Shell::Bash), "myapp", &mut std::io::stdout());
+    /// ```
+    pub fn from_env() -> Result<Shell, String> {
+        let env_shell = std::env::var_os("SHELL").map(PathBuf::from);
+        if let Some(shell) = env_shell
+            .as_ref()
+            .and_then(|s| s.file_name())
+            .and_then(|s| s.to_str())
+        {
+            Ok(shell.parse::<Shell>()?)
+        } else if cfg!(windows) {
+            // Assume powershell for windows
+            Ok(Shell::PowerShell)
+        } else {
+            Err("SHELL environment variable not set".to_string())
+        }
```

Comment:
This isn't quite true as it could be set to a value without a file name, to non-UTF8 filename, etc